src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,311 @@
+#    limitations under the License.
+
+"""A SAS-TH HTTP Server implementation, based on v1.2 of the SAS-SAS TS.
+
+A SAS-TH test server could be run by creating the object for the class SasTestHarnessServer
+and by invoking the API start()".
+"""
+
+import hashlib
+import inspect
+import json
+import logging
+import os.path
+import random
+import ssl
+import sys
+import threading
+import time
+from datetime import datetime, timedelta
+from urllib import quote
+from BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler
+
+CERT_FILE = 'certs/server.cert'
+KEY_FILE = 'certs/server.key'
+CA_CERT = 'certs/ca.cert'
+CIPHERS = [
+    'AES128-GCM-SHA256',              # TLS_RSA_WITH_AES_128_GCM_SHA256
+    'AES256-GCM-SHA384',              # TLS_RSA_WITH_AES_256_GCM_SHA384
+    'ECDHE-ECDSA-AES128-GCM-SHA256',  # TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+    'ECDHE-ECDSA-AES256-GCM-SHA384',  # TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+    'ECDHE-RSA-AES128-GCM-SHA256',    # TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+]
+
+
+class SasTestHarnessServer(threading.Thread):
+  """
+  Test Harness acting as a HTTP Server running in background and serving Pull/GET requests.
+  """
+  def __init__(self, name, base_url, dump_path, cert_file=None, key_file=None):
+    """ This constructor initializes the SasTestHarnessServer class with required information
+    passed as args
+
+    Args:
+        name: A name for SAS-TH.If multiple SAS-THs are created then this value should be unique.
+        base_url: base url of SAS-TH that the SAS UUT should connect to for FAD records.
+        dump_path: file_path where the FAD records served by the SAS-TH will be placed.
+        cert_file: The certificate used by SAS-TH to authorize with SAS UUT.
+        key_file:  The private key file for the cert_file.
+    """
+    super(SasTestHarnessServer, self).__init__()
+    self.name = name
+    self.http_server_url = base_url
+    self.dump_path = os.path.join(name, dump_path)
+    self.cert_file = cert_file if cert_file is not None else CERT_FILE
+    self.key_file = key_file if key_file is not None else KEY_FILE
+    self.daemon = True
+    self.server = None
+
+
+  def getSasThUrl(self):
+    return self.http_server_url
+
+  def getUrlHostNameAndPort(self):
+    """ This method will return the http_server_url used in startServer method by truncating other
+    querystring in url and returns tuple(<hostname>, <port>)
+    """
+    splitted_url = self.http_server_url.split('/')[2:]
+    logging.info(splitted_url)
+    return splitted_url[0].split(':')[0], int(splitted_url[0].split(':')[1])
+
+  def getDumpFilePath(self):
+    """ This method will return absolute path where FAD Records of SAS-TH are stored.
+    """
+    harness_dir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+    dumpfile_path = os.path.join(harness_dir, 'FADRecords', self.dump_path)
+    return dumpfile_path
+
+  def run(self):
+    self.startServer()
```

Comment:
The   run() method is an overridden method from thread class and it gets invoked   when the thread is started using thread.start().          The start server code has been moved into this method to avoid confusion.